nodeJS/express/using_postgresql.md

```diff
@@ -0,0 +1,309 @@
+  connectionString: "postgresql://<role_name>:<role_password>@localhost:5432/top_users"
+});
+```
+
+Pick whichever method you want to use and let's continue.
+
+<div class="lesson-note" markdown="1">
+
+#### Two ways of connecting with pg
+
+`pg` has two ways to connect to a db: a client and a pool.
+
+Client is an individual connection to the DB, which you manually manage. You open a connection, do your query, then close it. This is fine for one-off queries, but can become expensive if you're dealing with a lot of queries. Wouldn't this problem be alleviated if we could somehow hold onto a client? Yes!
+
+Enter pool. As the name suggests, it's a pool of clients. A pool holds onto connections. And when you query, it'll programmatically open a new connection unless there's an existing spare one. Perfect for web servers.
+
+</div>
+
+### Querying with pg
+
+With our initialized `Pool`, we can use the `query` method. Create a new `db/queries.js` file. Upon revising our project requirements, we understand we need two db interactions: getting all usernames and inserting a new username. Let's define these functions:
+
+```javascript
+const pool = require("./pool");
+
+async function getAllUsernames() {
+  const { rows } = await db.query("SELECT * FROM usernames");
+  return rows;
+}
+
+async function insertUsername(username) {
+  await db.query("INSERT INTO usernames (username) VALUES ($1)", [username]);
+}
+
+module.exports = {
+  getAllUsernames,
+  insertUsername
+}
+```
+
+<div class="lesson-note" markdown="1">
+
+#### Parameterization
+
+What's with the `$1` in the insert query?
+
+Alternatively, the query could look like:
+
+```javascript
+await db.query("INSERT INTO usernames (username) VALUES ('" + username + "')");
+```
+
+We're passing user entered value i.e. `username` directly into our query. A nefarious user could enter something like `sike'); DROP TABLE usernames; --` and wreak havoc. Scary stuff. This is called [SQL injection](https://en.wikipedia.org/wiki/SQL_injection).
+
+`pg` provides [query parameterization](https://node-postgres.com/features/queries#parameterized-query) to prevent this. Instead of passing user input directly, we pass it in an array as the second argument. `pg` handles the rest.
+
+</div>
+
+Invoke the above two functions in the specific controllers (you might have different function names etc. Important thing is to understand how the db functions are invoked):
+
+```javascript
+const db = require("../db/queries");
+
+async function getUsernames(req, res) {
+  const usernames = await db.getAllUsernames();
+  console.log("Usernames: ", usernames);
+  res.send("Usernames: " + usernames.map(user => user.username).join(", "));
+}
+
+async function createUsernameGet(req, res) {
+  // render the form
+}
+
+async function createUsernamePost(req, res) {
+  const { username } = req.body;
+  await db.insertUsername(username);
+  res.redirect("/");
+}
+
+module.exports {
+  getUsernames,
+  createUsername
+}
+```
```

Comment:
Since these are middleware that can throw, we should use `express-async-handler` as is introduced in the [Controllers lesson](https://github.com/TheOdinProject/curriculum/pull/27695).

Also fixing the exports bit

````suggestion
```javascript
const db = require("../db/queries");
const asyncHandler = require("express-async-handler");

const getUsernames = asyncHandler(async (req, res) => {
  const usernames = await db.getAllUsernames();
  console.log("Usernames: ", usernames);
  res.send("Usernames: " + usernames.map(user => user.username).join(", "));
});

const createUsernameGet = asyncHandler(async (req, res) => {
  // render the form
});

const createUsernamePost = asyncHandler(async (req, res) => {
  const { username } = req.body;
  await db.insertUsername(username);
  res.redirect("/");
});

module.exports = {
  getUsernames,
  createUsernameGet,
  createUsernamePost,
};
```
````

Reply:
I've skipped A LOT of things by design like error handling, data validation, and `express-async-handler`. This code example is just there to show how the learner might use their database query functions, that's it. The complete controller implementation is on them.

Reply:
That's fair enough then. We'd still need to fix the module.exports bit for the `=` and right function names

```js
module.exports = {
  getUsernames,
  createUsernameGet,
  createUsernamePost,
};
```

Reply:
Adding to the above, I'm not expecting learners to implement everything they've learned in the controllers lesson for the app they build this lesson. The purpose for this lesson's app is to practice db stuff only.

Reply:
fixed